integration_tests/src/main/python/parquet_test.py

```diff
@@ -779,6 +780,17 @@ def test_spark_32639(std_input_path):
         lambda spark: spark.read.schema(schema_str).parquet(data_path),
         conf=original_parquet_file_reader_conf)
 
+@pytest.mark.skipif(spark_version() >= '3.2.0' and spark_version() < '3.2.4', reason='Since 3.2.4')
+@pytest.mark.skipif(spark_version() >= '3.3.0' and spark_version() < '3.3.2', reason='Since 3.3.2')
+def test_spark_40819(std_input_path):
+    data_path = "%s/timestamp-nanos.parquet" % (std_input_path)
+    conf = copy_and_update(original_parquet_file_reader_conf, {
+            'spark.sql.legacy.parquet.nanosAsLong': True })
```

Comment:
We want to fallback to the CPU during tagging if this conf is true rather than crash the query.  If the conf is false, we want to throw the same exception Spark does if the nano timestamp is found in the Parquet schema.  See https://github.com/apache/spark/blob/v3.3.2/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L264-L269

Reply:
Yes to test the later one we would need to read the column, but with an explicit schema where we say it is a long instead of relying on schema discovery in Spark.

Reply:
I've added a test for the case where the conf is false. We already had the correct behavior in this case, but good to add the test for this.

In the tagging code, we don't currently have access to the Parquet logical types, so do we just want to fall back based on the config even if there is no timestamp with nano field in the file being read?

Reply:
Prior to Spark 3.2.4 it inadvertently "supported" nanos in Parquet. 

If we want to match Spark's behavior bug-for-bug, like in the past, then shouldn't we also support nanos to fall back to Longs prior to Spark 3.2.4? My question is why are we making an exception, in this case, to not match Spark?

Reply:
> In the tagging code, we don't currently have access to the Parquet logical types, so do we just want to fall back based on the config even if there is no timestamp with nano field in the file being read?

Yes, if we can't easily support loading nanos as longs.  If we can, then no need to do anything special for tagging, but we need tests showing it works properly.

> My question is why are we making an exception, in this case, to not match Spark?

The intent is to match Spark behavior, as long as we can do it correctly.  If we can load nanos as longs via cudf, then this should be straightforward.  However we do need the nanos check becomes an exception behavior in Spark 3.2.4 and other Sparks with the recent change, because Spark 3.2.4 will throw an exception on the Parquet nanos type by default.

Reply:
@andygrove can you please verify this? If cudf is doing the right thing then we should support nanos in Parquet by default prior to 3.2.4 and for later versions we should if the user has set the `nanosAsLong` conf

Reply:
I updated the PR description to outline the use cases I need to handle in this PR.

@razajafri I tried just following Spark and saying that we support long, but hit this error. 

```
E                   Caused by: java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=3, type=INT64, nullCount=Optional.empty, offHeap=(ID: 3 7f9580470880)}, ColumnVector{rows=3, type=STRING, nullCount=Optional.empty, offHeap=(ID: 4 7f9580471100)}, ColumnVector{rows=3, type=TIMESTAMP_MICROSECONDS, nullCount=Optional.empty, offHeap=(ID: 5 7f9580470ed0)}], cudfTable=140280079000480, rows=3} to [LongType, StringType, LongType] columns 0 to 3
E                       at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:789)
E                       at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:670)
E                       at com.nvidia.spark.rapids.CachedGpuBatchIterator$.$anonfun$apply$2(GpuDataProducer.scala:163)
```

Reply:
It's blowing up because cudf is sending it back as a Timestamp_microseconds. It should be a long for 3.1.x until 3.2. Then for every version without the conf it should blow up with an exception otherwise it should honor the conf and behave either as 3.1.x or as 3.2